app/src/main/java/org/openobservatory/ooniprobe/activity/BrowserActivity.java

```diff
@@ -0,0 +1,254 @@
+        webView.getSettings().setLoadWithOverviewMode(true);
+        webView.getSettings().setUseWideViewPort(true);
+        LoadWebViewUrl(urls.get(urlIndex));
+    }
+
+    private void setListeners() {
+        back.setOnClickListener(this);
+        forward.setOnClickListener(this);
+        refresh.setOnClickListener(this);
+        close.setOnClickListener(this);
+        share.setOnClickListener(this);
+        try_mirror.setOnClickListener(this);
+    }
+
+    void reloadButtons(){
+        if (webView.canGoBack()){
+            back.setAlpha(.5f);
+            back.setClickable(false);
+        }
+        else {
+            back.setAlpha(.5f);
+            back.setClickable(false);
+        }
+        if (webView.canGoForward()){
+            forward.setAlpha(.5f);
+            forward.setClickable(false);
+        }
+        else {
+            forward.setAlpha(.5f);
+            forward.setClickable(false);
+        }
+    }
+
+    @Override
+    public void onClick(View view) {
+        switch (view.getId()) {
+            case R.id.webviewBack:
+                isWebViewCanGoBack();
+                break;
+            case R.id.webviewForward:
+                if (webView.canGoForward())
+                    webView.goForward();
+                break;
+            case R.id.webviewReload:
+                String url = webView.getUrl();
+                LoadWebViewUrl(url);
+                break;
+            case R.id.webviewClose:
+                finish();
+                break;
+            case R.id.shareButton:
+                shareUrl();
+                break;
+            case R.id.tryMirror:
+                tryMirror();
+                break;
+        }
+    }
+
+
+    private class MyWebViewClient extends WebViewClient {
+        @Override
+        public boolean shouldOverrideUrlLoading(WebView view, String url) {
+            view.loadUrl(url);
+            return true;
+
+        }
+
+        @Override
+        public void onPageStarted(WebView view, String url, Bitmap favicon) {
+            super.onPageStarted(view, url, favicon);
+            refresh.setVisibility(View.GONE);
+            urlLabel.setText(getString(R.string.loading));
+            if (!webViewProgressBar.isShown())
+                webViewProgressBar.setVisibility(View.VISIBLE);
+        }
+
+        @Override
+        public void onPageFinished(WebView view, String url) {
+            super.onPageFinished(view, url);
+            refresh.setVisibility(View.VISIBLE);
+            if (webViewProgressBar.isShown())
+                webViewProgressBar.setVisibility(View.GONE);
+            System.out.println("Finished loading " + url);
+            if (url.substring(0, 5).equals("https"))
+                urlLabel.setText("\uD83D\uDD12 " + url);
+            else
+                urlLabel.setText(url);
+            reloadButtons();
+        }
+
+        //TODO keep the error toasts for all messages? are they necessary?
```

Comment:
@bassosimone @hellais what do you think about the next three functions? should I keep the Toast?